api/src/controllers/search.js

```diff
@@ -291,6 +258,89 @@ exports.protectedGet = function (args, res, next) {
   executeQuery(args, res, next);
 };
 
+// Generates the main match query, and optionally generates the master field match to be used
+// later in the pipeline.
+var generateMatchesForAggregation = async function (and, or, searchProperties, properties, schemaName, roles) {
+  // query modifiers
+  // Pluck the __master elements, and the flavour elements.  process them in different parts of the
+  // pipeline because of the linking of flavour to master records.
+  let __flavour = {};
+  let __master = {};
+  for (item in and) {
+    if ( item.startsWith('_master.')) {
+      __master[item] = and[item];
+    } else {
+      __flavour[item] = and[item];
+    }
+  }
+
+  defaultLog.info("__master:", __master);
+  defaultLog.info("__flavour:", __flavour);
+
+  const andExpArray = await generateExpArray(__flavour) || [];
+  const andMasterExpArray = await generateExpArray(__master, '_master.') || [];
+
+  defaultLog.info("andExpArray:", andExpArray);
+  defaultLog.info("andMasterExpArray:", andMasterExpArray);
+
+  // filters
+  // query modifiers
+  // Pluck the __master elements, and the flavour elements.  process them in different parts of the
+  // pipeline because of the linking of flavour to master records.
+  let __flavourOr = {};
+  let __masterOr = {};
+  for (item in or) {
+    if ( item.startsWith('_master.')) {
+      __masterOr[item] = or[item];
+    } else {
+      __flavourOr[item] = or[item];
+    }
+  }
+
+  defaultLog.info("__masterOr:", __masterOr);
+  defaultLog.info("__flavourOr:", __flavourOr);
+
+  const orExpArray = await generateExpArray(__flavourOr) || [];
+  const orMasterExpArray = await generateExpArray(__masterOr, '_master.') || [];
+
+  defaultLog.info("orExpArray:", orExpArray);
+  defaultLog.info("orMasterExpArray:", orMasterExpArray);
+
+  var modifier = {};
+  if (andExpArray.length > 0 && orExpArray.length > 0) {
+    modifier = { $and: [{ $and: andExpArray }, { $and: orExpArray }] };
+  } else if (andExpArray.length === 0 && orExpArray.length > 0) {
+    modifier = { $and: orExpArray };
+  } else if (andExpArray.length > 0 && orExpArray.length === 0) {
+    modifier = { $and: andExpArray };
+  }
+
+  var masterModifier = {};
+  if (andMasterExpArray.length > 0 && orMasterExpArray.length > 0) {
+    masterModifier = { $and: [{ $and: andMasterExpArray }, { $and: orMasterExpArray }] };
+  } else if (andMasterExpArray.length === 0 && orMasterExpArray.length > 0) {
+    masterModifier = { $and: orMasterExpArray };
+  } else if (andMasterExpArray.length > 0 && orMasterExpArray.length === 0) {
+    masterModifier = { $and: andMasterExpArray };
+  }
```

Comment:
Shouldn't at least some of these use $or ?

Reply:
The logic is that you $and the split elements coming from the API call, and if they are arrays, you'll see in the generate EXP they get constructed as an $or.  It should all work.